call MIDI input script functions with appropriate "this" object

Be-ing opened the pull request:
Fixing bug https://bugs.launchpad.net/mixxx/+bug/1567203

Be-ing committed:
call MIDI input script functions with appropriate "this" object

Pegasus-RPG commented:
Won't this cause performance problems if a script makes liberal use of the "this" keyword?

Be-ing commented:
Wouldn't the use of m_scriptValueCache in [ControllerEngine::resolveFunction](https://github.com/mixxxdj/mixxx/blob/master/src/controllers/controllerengine.cpp#L107) alleviate that?

Be-ing commented:
Do you think it would make sense to rename resolveFunction() to resolveObject()?

daschuer commented:
Thank you for working on this.

I think we need to to some more refactoring to not introduce a performance regression. 

It looks like for your approach ControllerEngine::resolveFunction is broken. 
How about return function and Object as output parameter? The Object and the function should be both cached in the m_scriptValueCache.

It looks like resolveFunction should return m_pEngine->globalObject() , if there is no obj found in the "obj.func" schema.

Be-ing commented:
What is the performance concern with the way it is? Making two function calls rather than one? How would you recommend returning two values, with a std::pair? tuple? QHash?

rryan commented:
> Won't this cause performance problems if a script makes liberal use of the "this" keyword? Or is resolveThisObject() intended to be called just during initialization?

The function body doesn't matter -- the only performance concern is that it adds an additional environment inspection per incoming MIDI message.

daschuer commented:
resolveFunction() might be longer than the final script call by dimensions. 
All QString function calls shoulder be reduced to a minimum. 
Since the can be very short we should squeeze the lookup code as much as possible.

daschuer commented:
Don't use a std::pair, tuple or similar anonymous structs. These are only useful for obfuscation. 

In this case you may consider to call the execute from inside the resolveFunction or use output parameters.

```
void resolveFunction(const QString& function,  QScriptValue* pThisObject, QScriptValue* pFunction ) const;
 ```

rryan commented:
Thanks for working on this! I've also been bothered by the right way to get 'this' pointing to something that makes sense in script handlers.

I think eventually we want to get away from resolveFunction and instead eval whatever expression the XML gave us. If the result of that eval is a function then we call/cache it. Doing dot-based environment traversal is kind of doing the work of eval but probably inaccurately. The "function prefixes" we use for init/shutdown are particularly hacky too and I'd like to see them go.

I think we might also encourage script writers to properly bind their event handlers to the object the handler lives in using Function.prototype.bind. We could even provide a helper method to bind handlers in our common Controller class we provide. Then it wouldn't matter what context we execute the method with. 

Anyhow.. what we do is roughly the equivalent of:

```
function Controller() {
  this.name = "FancyController";
}

var MyController = new Controller();
MyController.handler = function() {
  console.log(this.name);
}
```

Mixxx then does the rough equivalent of:
```js
var theHandler = eval("MyController.handler");
theHandler();
```

And you get:
```
Cannot read property 'name' of undefined
```

OTOH if you do:

```js
eval("MyController.handler()");
```
You get"
```
FancyController
```

If the function were bound:

```js
MyController.handler = function() {
  console.log(this.name);
}
MyController.handler = MyController.handler.bind(MyController);

var theHandler = eval("MyController.handler");
theHandler();
```

Then it would work as expected:
```
FancyController
```

Be-ing commented:
> All QString function calls shoulder be reduced to a minimum. 

How about storing the QScriptValue for the "this" object for each function name in a cache along with the QScriptValue for the function?

Be-ing commented:
> I think we might also encourage script writers to properly bind their event handlers to the object the handler lives in using Function.prototype.bind. We could even provide a helper method to bind handlers in our common Controller class we provide. Then it wouldn't matter what context we execute the method with. 

I'd rather that Mixxx found the appropriate "this" object itself. My broader goal is to make writing complex scripts more convenient with a minimum of repetitive boilerplate code. IMO binding every callback function is annoying. I can test again, but I think I did try binding a callback function but it didn't do anything (probably because the C++ called the function with `this` explicitly set to the global object anyway).

Be-ing commented:
> I think eventually we want to get away from resolveFunction and instead eval whatever expression the XML gave us. If the result of that eval is a function then we call/cache it. Doing dot-based environment traversal is kind of doing the work of eval but probably inaccurately. The "function prefixes" we use for init/shutdown are particularly hacky too and I'd like to see them go.

I agree. I think this could also resolve https://bugs.launchpad.net/mixxx/+bug/1565377

daschuer commented:
Yes, we should do this. 
It would be even better to evaluate the strings only once and store the resolved script values along with the strings. Don't know if this can be part of this PR or not.  

We have actually real performance issues with High resolution jog wheels here: 
For example the SCS.1d uses 3xmidi speed which results in an update rate of 
3125 messages per second

Be-ing commented:
> We have actually real performance issues with High resolution jog wheels here:
For example the SCS.1d uses 3xmidi speed which results in an update rate of
3125 messages per second

Yes, I am aware. I do not have any SCS controllers or controllers with jog wheels though, so others would have to test if changes impact real usability.

Be-ing commented:
Hm, I realized that it's kinda silly to handle this upon receipt of every MIDI message. It would make more sense to store the QScriptValues in a cache upon loading the mapping, then refer to that when MIDI messages are received.

daschuer commented:
We do not need a SCS to check the performance changes. 
Just counting the function calls parsing the string will give a good estimation. 
To proof it, you may set up a ScopedTimer like we did in other places. 

It is obvious that every "." in the sting is an extra cache lookup. 
If we can replace the cache, by a storage along with the string, it allows us to use things like
P32.leftDeck.hotcueButtons[2].input with less pain.

rryan commented:
Here's how the standard says this should work:
http://www.ecma-international.org/ecma-262/5.1/#sec-11.2.3
http://www.ecma-international.org/ecma-262/5.1/#sec-8.7

So this describes why 
```js
foo.bar(1, 2, 3)
```
calls bar with this set to foo but 
```js
var f = foo.bar;
f(1, 2, 3);
```
calls bar with this set to the global object. 

The grammar production rule for function calls distinguishes between an object reference and a free standing value. When foo.bar is part of the expression bar is an object reference but when it's a standalone value it's not.

Our traversal of the environment and call of bar is equivalent to the latter example. To get normal javascript behavior, we probably want the former.

Using the prior value in the dot traversal would mostly approximate the GetBase function defined by the standard for object references so this is a pretty reasonable fix! However it wouldn't work for 
```js
foo.bar[foo.currentDeck].handler
``` 
For that we probably need to build a Javascript expression parser to do it right (notice how splitting on dots no longer works for that case).

rryan commented:
Also note that if we want to support:
```js
foo.bar[foo.currentDeck].handler
```
this means caching is no longer an option since we need to re-evaluate foo.currentDeck each time to be correct

rryan commented on src/controllers/midi/midicontroller.cpp:
```diff
@@ -289,7 +289,8 @@ void MidiController::processInputMapping(const MidiInputMapping& mapping,
 
         QScriptValue function = pEngine->resolveFunction(mapping.control.item);
-        if (!pEngine->execute(function, channel, control, value, status,
+        QScriptValue thisObject = pEngine->resolveFunction(mapping.control.item.section(".", 0, -2));
```
please also update sysex handling (line 551) for consistency

rryan commented:
So I would be in favor of merging this as a short term fix. To minimize perf impact, resolveFunction and the script value cache should probably be updated to return a std::pair of (context, value) where context is the last value in the dot chain before value -- that way we'd be doing essentially no more work than we already are per MIDI message.

Be-ing commented:
Okay, I'll work on that as a short-term fix. I think trying to implement fully correct JS behavior by parsing strings in C++ would likely end in other hacks that don't really work with the full flexibility of JS. I think to really take advantage of all JS has to offer, the input handler callbacks need to be registered by scripts (generally in their init functions).

daschuer commented:
Ok, I understand we cannot cache the object because it may vary. 

But `foo.bar[foo.currentDeck].handler` is anyway too complex to compile it from a string on each midi update. 

I think there are some possible ways to solve it:
* disallow such expressions where the "this" context varies. 
* detect the case (multi dots or arrays) and wrap it into an anonymous function 

Pseudo code:
```
foo.mapping8044 = function (channel, control, value, status, group) {
    foo.bar[foo.currentDeck].handler(channel, control, value, status, group);
}
```

```
QString wrapper = 
        "foo.mapping8044 = function (channel, control, value, status, group) {" +
        mapping.control.item +
        " (channel, control, value, status, group); }";
m_pEngine->evaluate(wrapper); 

// resove foo
// resove mapping8044
// Store both QScriptValues along with "mapping.control.item"
 ```
If we found out the compiler is smart enough to ditch the extra call, we can always use the evaluate. 
We may also allow to inline whole script functions to the xml file.  

-----

> should probably be updated to return a std::pair 

never use std::pair is a source of later confusion. These are good alternatives
* output parameters 
* define your own struct / class

rryan commented:
> never use std::pair is a source of later confusion. These are good alternatives
> 
> output parameters
> define your own struct / class

Hm, I disagree with "never use std::pair". It's not too hard to read the function documentation to learn what it returns -- it's the same as reading the function prototype to learn what the output parameters are or reading the struct definition to learn what members it has. I agree that a pair isn't good to use as a data model that is prevalent across the codebase but for an internal implementation detail it's fine.

rryan commented:
> wrap it into an anonymous function
> ...
> If we found out the compiler is smart enough to ditch the extra call, we can always use the evaluate. 

Unless we have data showing an extra pre-compiled-by-the-interpreter function call is worse than what we have now with resolveFunction poking at the environment repeatedly from the QtScript API then I wouldn't worry about it. Javascript interpreters are *fast* from years of optimizations trying to get fancy webapps to work faster and I'm not sure our QtScript API and string manipulation would win against a JS interpreter's evaluation of the user provided expression -- especially if the code is already parsed (which it would be after our one-time evaluate) and better yet it would be a correct implementation of the JS spec! 

On the whole wrapping every script handler in an anonymous function is a pretty nice solution and only requires the string formatting once up front.

Pro:
* we only have to evaluate the wrapper once -- no need for resolveFunction
* the variables pointed to by the expression provided by the user can change or not.
* complicated expressions work as long as they return a function:  ```MyController.deck[MyController.activeDeck].Handler("hotcue_10")``` is valid if it returns a function.
* we get the "real" Javascript foo.bar() behavior where bar is executed with this="foo". 

Con:
* None demonstrated so far?

daschuer commented:
There are situations where pair helps, but they are very rare. You can use std::tie to make it less scary. But even that involves some anonymous boiler plate which should be avoided. 

In this case this reads well: 
```
QScriptValue thisObject; 
QScriptValue function;
resolveFunction(mapping.control.item, &thisObject, &function );
if (!pEngine->execute(function, thisObject, channel, control, value, status,
                              mapping.control.group, timestamp)) {
```
If the pointers are to scary for you we may use 

```
struct ScriptMemberFunction{
    QScriptValue thisObject; 
    QScriptValue function;
}

ScriptMemberFunction memberFunction = resolveFunction(mapping.control.item, pThisObject, pFunction );
if (!pEngine->execute(memberFunction.function, memberFunction.thisObject, channel, control, value, status,
                              mapping.control.group, timestamp)) {
```

This is far more self explaining than 

```
std::pair<QScriptValue, QScriptValue> memberFunction = resolveFunction(mapping.control.item, pThisObject, pFunction );
if (!pEngine->execute(memberFunction.fist, memberFunction.second, channel, control, value, status,
                              mapping.control.group, timestamp)) {
```

rryan commented:
So here's my proposal based on @daschuer's -- at script parse time turn every handler into an anonymous function and evaluate it once then and not when MIDI messages come in:
```c++
QString wrapper = "function (channel, control, value, status, group) { (" + mapping.control.item + ")(channel, control,value, status, group); }";
m_pEngine->evaluate(wrapper); 
```

rryan commented:
> There are situations where pair helps, but they are very rare. You can use std::tie to make it less scary. But even that involves some anonymous boiler plate which should be avoided.

I agree that if the type will spread beyond the file it lives in then it's best to define a type for it. However I think we're going to have to agree to disagree on whether it's ok to use std::pair in local contexts.

Be-ing commented:
I think that wrapper will need a bit of modification:
`QString wrapper = "function (channel, control, value, status, group) { (" + 
    mapping.control.item + ").apply(this, arguments); }";
m_pEngine->evaluate(wrapper); `

Be-ing commented:
Using an anonymous wrapper function could help for something simple like `MyController.leftDeck.hotcues[1]`, but as soon as there are any more dots, there will still be the issue of how to get `this`.

rryan commented:
> I think that wrapper will need a bit of modification:
> 
> QString wrapper = "function (channel, control, value, status, group) { (" + 
>     mapping.control.item + ").apply(this, arguments); }";
> m_pEngine->evaluate(wrapper);

In the proposal I'm making, ```this``` would always be the global object we invoke these anonymous functions with so using apply wouldn't change anything here.

> Using an anonymous wrapper function could help for something simple like MyController.leftDeck.hotcues[1], but as soon as there are any more dots (like MyController.deck[MyController.activeDeck].Handler("hotcue_10")), there will still be the issue of how to get this.

I'm proposing Mixxx get out of the business of mucking around with a specific this. We invoke the expression you provided us and you get what Javascript gives you as per the spec.

For ```MyController.deck[MyController.activeDeck].Handler("hotcue_10"))```, ```Handler```'s this will be the result of ```deck[MyController.activeDeck]``` and the script writer will have to take care to bind the handler they return (as they would in a regular web app or other context of using javascript).

However for something more sane like: ```MyController.deck[MyController.activeDeck].hotcueHandler```

hotcueHandler's this will be the current active deck object without the script writer having to bind anything.

The most common case of: ```MyController.hotcueHandler``` would call hotcueHandler with this equal to MyController.

daschuer commented:
sorry, the "Close and comment" button strikes me

Be-ing commented:
Ah, I see. Of course the wrapped function's `this` will be correct when it's evaluated from any JS context. That obviates the need to keep both the function's QScriptValue and the thisObject's QScriptValue in a cache.

rryan commented:
Here's a little example: https://repl.it/CDxj/2

Be-ing commented:
Hm, I'm getting a kinda confusing error. The C++ I tried:
```c++
        QString wrappedCode = "(function (channel, control, value, status, group) { (" +
                              codeSnippet + ")(channel, control, value, status, group); })";
        wrappedFunction = evaluate(wrappedCode);
```
Mixxx error:
```
Debug [Controller]: "Hercules P32 DJ MIDI 1: t:115208 ms status 0x91 (ch 2, opcode 0x9), ctrl 0x0A, val 0x7F" 
Debug [Controller]: ControllerEngine: Loading "(function (channel, control, value, status, group) { (P32.leftDeck.play.input)(channel, control, value, status, group); })" 
Warning [Controller]: "ControllerEngine: Problem opening the script file: (function (channel, control, value, status, group) { (P32.leftDeck.play.input)(channel, control, value, status, group); }), error # 5, No such file or directory" 
Debug [Controller]: ControllerEngine::internalExecute: QVariant(bool, false) Not a function 
Debug [Controller]: MidiController: Invalid script function "P32.leftDeck.play.input"
```
I first tried the code @rryan suggested without the parentheses wrapping all the JS code and got the same error. I checked the QScriptEngine documentation and saw this example:
```c++
QScriptValue fun = myEngine.evaluate("(function(a, b) { return a + b; })");
QScriptValueList args;
args << 1 << 2;
QScriptValue threeAgain = fun.call(QScriptValue(), args);
```
so I don't know why this isn't working.

Be-ing commented:
Aha, nevermind. I realized there's a method of ControllerEngine called `evaluate` that I was calling that isn't QScriptEngine::evaluate.

Be-ing committed:
Revert "call MIDI input script functions with appropriate "this" object"

This reverts commit 473297046cd8eca24bada519c10c9be49fbd5301.

Be-ing committed:
treat <key> fields in MIDI mapping XML files for script functions as JS code

by wrapping in anonymous functions. This resolves these bugs:
https://bugs.launchpad.net/mixxx/+bug/1567203
https://bugs.launchpad.net/mixxx/+bug/1565377

Be-ing commented:
Okay, I implemented the wrapping in an anonymous function. `this` is assigned correctly and arrays can be used in the XML.

rryan commented on src/controllers/controller.cpp:
```diff
@@ -121,7 +121,7 @@ void Controller::receive(const QByteArray data, mixxx::Duration timestamp) {
         }
         function.append(".incomingData");
-        QScriptValue incomingData = m_pEngine->resolveFunction(function);
+        QScriptValue incomingData = m_pEngine->wrapFunctionCode(function);
```
incomingData functions take 2 arguments -- not the 5 MIDI handlers take

Maybe you could add a "num arguments" to wrapFunctionCode and then create the argument list with a loop (the names of the arguments dont matter so you could make them like 'arg1', 'arg2', 'arg3', etc.)

rryan commented on src/controllers/controllerengine.cpp:
```diff
@@ -98,30 +98,44 @@ void ControllerEngine::callFunctionOnObjects(QList<QString> scriptFunctionPrefix
+        wrappedFunction = i.value();
+    } else {
+        QScriptValue function = m_pEngine->evaluate(codeSnippet);
+        if (!syntaxIsValid(codeSnippet)) {
```
May as well check the syntax before evaluating

rryan commented on src/controllers/controllerengine.cpp:
```diff
@@ -98,30 +98,44 @@ void ControllerEngine::callFunctionOnObjects(QList<QString> scriptFunctionPrefix
+            return m_pEngine->evaluate("(function () {})");
         }
+
+        m_scriptWrappedFunctionCache[codeSnippet] = wrappedFunction;
```
resolveFunction cached failed lookups but since we return early the next time this MIDI messages comes in we'll repeat the syntax check / evaluate / failure dialog, etc. I think we should probably cache failures too here. WDYT?

Be-ing replied:
I thought about that, but I realized it could be possible for a failed lookup to work the next time if it references script variables that the script has changed between calls.

rryan replied:
Gotcha -- I could be wrong but I think evaluating the wrapper can only fail if there is a syntax error -- for example: https://repl.it/CEMr

Undefined variables or bad references would be exceptions at the time we call the wrapper function.

And I can't think of a syntax error you can write that will become not a syntax error based on changed in the environment.

Be-ing replied:
Evaluating the wrapper fails when the passed code snippet is a valid object but not a function. The script could reassign the object to a function between MIDI messages received.

rryan replied:
Huh, that doesn't match my understanding. For example: https://repl.it/CENd
Evaluating the wrapper succeeds without a problem. Could you post an example of this?

Be-ing replied:
https://repl.it/CENh/1

Map some button's input to P32.button[1] and another's to P32.button[2]. If button 1 is pressed first, evaluating the wrapper would throw an exception. If button 2 is pressed before button 1 is pressed again, pressing button 1 should work.

Be-ing replied:
This is the error I get from Mixxx:
```Uncaught exception at line 1 in passed code: TypeError: Result of expression '(P32.whatever)' [stuff] is not a function.
Backtrace:
<anonymous>()@:1```
P32.whatever is set to the string `'stuff'`.

rryan replied:
Hm, are you sure that error occurred in evaluating the wrapper function or was it executing the wrapper function? 

I wrote a quick test that suggests evaluating the wrapper would work fine in that case:
https://github.com/rryan/mixxx/commit/7036720de8064e273e0ab0dd9286cf19586c21c5

Output:
```
[ RUN      ] ControllerEngineTest.Evaluate
Loading resources from  "/Users/rjryan/Code/mixxx/res/"
ConfigObject: Could not read "/Users/rjryan/Code/mixxx/src/test/test_data/test.cfg"
SetUp
ControllerEngine: Loading "/var/folders/qr/klrx2z4s0l37vkk8y5lq5718002y8v/T/qt_temp.M56742"
TypeError: Result of expression '(P32.whatever)' [stuff] is not a function.
whatever works!
TearDown
ControllerEngine shutting down...
Warning: Test leaked control: "[Test]" "potmeter_toggle"
Warning: Test leaked control: "[Test]" "potmeter_up_small"
Warning: Test leaked control: "[Test]" "potmeter_set_default"
Warning: Test leaked control: "[Test]" "potmeter_down_small"
Warning: Test leaked control: "[Test]" "potmeter_set_one"
Warning: Test leaked control: "[Test]" "potmeter_set_minus_one"
Warning: Test leaked control: "[Test]" "potmeter_down"
Warning: Test leaked control: "[Test]" "potmeter_minus_toggle"
Warning: Test leaked control: "[Test]" "potmeter_set_zero"
Warning: Test leaked control: "[Test]" "potmeter"
Warning: Test leaked control: "[Test]" "potmeter_up"
[       OK ] ControllerEngineTest.Evaluate (1 ms)

```

Be-ing replied:
Ah, you are correct. I misunderstood where the exception was occuring. It is indeed when calling the function in internalExecute(). So yes, it does make sense to always cache the function.

rryan commented:
Thanks! This so much nicer than what we had before. 

It would be nice to have a test in ControllerEngineTest that checks this :).

Be-ing commented:
> It would be nice to have a test in ControllerEngineTest that checks this :).

It would. It would also be nice if ControllerEngineTest.scriptConnectDisconnectControlNamedFunction didn't segfault (which was not caused by the change this PR makes).

rryan commented:
> It would. It would also be nice if ControllerEngineTest.scriptConnectDisconnectControlNamedFunction didn't segfault (which was not caused by the change this PR makes).

Yikes, on your machine? It doesn't for me and Travis passed for both Linux and Mac which means it didn't there either. Could you grab a backtrace?

Be-ing committed:
add numberOfArgs argument to ControllerEngine::wrapFunctionCode

Be-ing commented:
Backtrace:
```
[       OK ] ControllerEngineTest.scriptGetSetValue (2 ms)
[ RUN      ] ControllerEngineTest.scriptConnectDisconnectControlNamedFunction
Loading resources from  "/home/be/local/share/mixxx/" 
ConfigObject: Could not read "/home/be/sw/mixxx/src/test/test_data/test.cfg" 
SetUp 
ControllerEngine: Loading "/tmp/qt_temp.X19276" 
Connection: "[Test]" "potmeter" 

Program received signal SIGSEGV, Segmentation fault.
0x00007ffff47805da in QString::fromLocal8Bit (str=0x21 <error: Cannot access memory at address 0x21>, size=size@entry=-1) at tools/qstring.cpp:3960
3960        if (size == 0 || (!*str && size < 0))
```

rryan commented:
> Backtrace:

Could you type ```thread apply all bt``` once it crashes to get the full stack trace? That's just the line it crashed on.

Be-ing commented:
Full backtrace attached
[test-crash.txt](https://github.com/mixxxdj/mixxx/files/212231/test-crash.txt)

Be-ing commented:
I'm using qt-4.8.7-12.fc23.x86_64

Be-ing committed:
always cache function in ControllerEngine::wrapFunctionCode

Also, print error messages through JS functions

Be-ing commented:
Okay, wrapFunctionCode() always caches the QScriptValue function now.

daschuer commented on src/controllers/controller.cpp:
```diff
@@ -121,7 +121,7 @@ void Controller::receive(const QByteArray data, mixxx::Duration timestamp) {
         }
         function.append(".incomingData");
-        QScriptValue incomingData = m_pEngine->resolveFunction(function);
+        QScriptValue incomingData = m_pEngine->wrapFunctionCode(function, 2);
```
This loop should be moved to: 
ControllerEngine::initializeScripts and fill a simple QList<QScripValue> m_incommongDataFunctions 
which is only called once. 

Here we can use than a simple .. without any string arithmetic: 
```
for (const auto& incomingData: m_pEngine->m_incommongDataFunctions()) {
        if (!m_pEngine->execute(incomingData, data, timestamp)) {
            qWarning() << "Controller: Invalid script function" << function;
        }
    }
```

daschuer commented on src/controllers/controllerengine.cpp:
```diff
@@ -98,30 +98,48 @@ void ControllerEngine::callFunctionOnObjects(QList<QString> scriptFunctionPrefix
+        if (!syntaxIsValid(codeSnippet)) {
+            wrappedFunction = m_pEngine->evaluate(
+                "(function () { print('Syntax error in: "
+                + codeSnippet + "'); })");
```
Cool! QScriptValue ControllerEngine::connectControl should also make use of this nice new function.

Be-ing replied:
There is no need for that. connectControl() already assigns the appropriate `this` object to the callback.

Be-ing replied:
I'm kinda afraid to touch the convoluted logic of that function after #656 . Before opening this PR I started working on [another branch](https://github.com/Be-ing/mixxx/tree/script_connection_object_extensions) to introduce a cleaner API for the undocumented functionality in that function.

Be-ing replied:
I don't understand why the capability to pass functions wrapped in strings was introduced in the first place. IMO it makes the C++ side unnecessarily complex, which led to the bug that inspired me to start #656. I think new functionality dealing with JS functions should only accept JS functions as input and not deal with the complexity of turning strings into functions.

Be-ing replied:
Like @rryan said about Mixxx getting out of the business of resolving the `this` object by hackishly parsing strings, I think Mixxx should move away from the business of handling strings of JavaScript at all. I would prefer to move towards a fully JS mapping option in which scripts manage their own input handler callbacks dynamically (as a side benefit mapping development would be nicer because the XML file would never have to be reloaded).

daschuer replied:
The embedded js code as strings works like a poor mans Lamda expression. This works quite nice for timer events. How will a js compliment solution look like?

daschuer replied:
I am quite pleased with the current solution with invoking js code from a XML mapping. Once we get rid of the extra hash lookup for the anonymous function, I can't think of a pure js solution that performs better.

A similar fast js solution will probably have a similar midi message lookup register that we have in c++ right now. Unfortunately we will ether loose the script fee direct Co mappings, or we have to live with a double lookup.

Be-ing replied:
My motivation isn't improving performance, it's improving flexibility in how scripts are structured. I'd imagine that the implementation would be similar to ControllerEngineConnectionScriptValue.

daschuer commented on src/controllers/controllerengine.cpp:
```diff
@@ -330,6 +341,25 @@ bool ControllerEngine::internalExecute(QScriptValue thisObject,
+    if (m_pEngine == nullptr)
+        return false;
+
+    if (!syntaxIsValid(scriptCode)) {
```
This call along with the wrapper generation should be done earlier, just after evaluating the QScriptValue as string places like: 
ControllerEngine::beginTimer

daschuer commented on src/controllers/midi/midicontroller.cpp:
```diff
@@ -288,7 +288,7 @@ void MidiController::processInputMapping(const MidiInputMapping& mapping,
         }
 
-        QScriptValue function = pEngine->resolveFunction(mapping.control.item);
+        QScriptValue function = pEngine->wrapFunctionCode(mapping.control.item, 5);
```
This call should be moved to 
void MidiController::learnTemporaryInputMappings and store result to a new mapping element.

daschuer replied:
And it looks like we have to fill this new element somewhere after MidiController::visit().

rryan replied:
IMO this would violate abstraction barriers and instead needs some somewhat involved refactoring to gain the performance wins you're talking about. (Script values escaping the ControllerEngine as they did with resolveFunction was already an abstraction violation and storing the values in the MidiController/preset is taking it a step further :) ).

For example -- we need to invalidate the wrapper script values when the filesystem watcher notices the script has changed. If the QScriptValue was stored in the MidiInputMapping (or an m_incomingDataFunctions list) there is no current easy way for the script engine to reach out into the MidiMapping/MidiController to do this. I'm not even sure if it's safe to use a QScriptValue after its QScriptEngine has been deleted (gracefulShutdown destroys the current engine).

daschuer replied:
OK, the script value holds a pointer of the script engine and vice versa. So it can be dangerous to spread QStringValues around. 

How about turn the current cache solution into a shared pointer cache. The mapping can hold a weak pointer. 
Turing a weak pointer into a shared pointer and execute the QScriptValue function should be much faster then looking up a hash. If the strong cache is gone, all stored weak pointers are invalid and the mapping has to evaluate its functions again. 

```
QHash<QString, QSharedPointer<QScriptValue> > m_scriptWrappedFunctionCache
...
QSharedPointer<QScriptValue> = pFunction(new QScriptValue(m_pEngine->evaluate(codeSnippet)));
m_scriptWrappedFunctionCache[codeSnippet] = pFunction; 
...

...
QWeakPointer<QScriptValue> scriptfunction;
...
mapping.scriptfunction = pFunction; 
...

QSharedPointer<QScriptValue> strong = mapping.scriptfunction.toStrongRef();
if (strong)
    // execute
else
   // evaluate again 
```

rryan replied:
That works -- though it creates more complexity / tight coupling between the controller and the script engine. We could create an opaque reference/token or something.

Unless @Be-ing wants to go for "extra credit" here from my POV this branch is a strict improvement over master though.

Be-ing replied:
I'd prefer to merge this sooner and move on to other projects at this point.

daschuer replied:
Yes, right. This brach includes already enough improvements to be mergable. Well done.  :-)

rryan commented on src/controllers/controllerengine.cpp:
```diff
@@ -98,30 +98,48 @@ void ControllerEngine::callFunctionOnObjects(QList<QString> scriptFunctionPrefix
+    if (i != m_scriptWrappedFunctionCache.end()) {
+        wrappedFunction = i.value();
+    } else {
+        QScriptValue function = m_pEngine->evaluate(codeSnippet);
```
I think we have some extra work being done here:
* evaluate codeSnippet (also does syntax checking but doesn't expose it to us)
* syntax check codeSnippet 
* evaluate wrappedFunction (syntax checks again)

Evaluating codeSnippet may fail for benign reasons (like you've pointed out, the user's expression may not evaluate successfully if objects aren't defined) -- so I think we should drop the evaluate(codeSnippet) entirely. 

Syntax checking codeSnippet is nice for getting a cleaner error message. Do we need it though? I wonder if the error message from checkException() after evaluating wrappedFunction would have a good enough error message?

Be-ing replied:
> Syntax checking codeSnippet is nice for getting a cleaner error message. Do we need it though? I wonder if the error message from checkException() after evaluating wrappedFunction would have a good enough error message?

It doesn't. It just gives a generic "Parse error" message without showing the source code that triggered the error. That's why I included the syntax check separate from the evaluation.

Be-ing replied:
The extra evaluation done on code with valid syntax is annoying and necessary because of how the QtScript API is. I think it's fine though because the results are cached.

rryan replied:
Gotcha -- so we can drop the evaluate(codeSnippet) but keep the syntax check then?

Be-ing replied:
This would require another nested `if ... else`` statement to avoid the extra evaluation in the case of a syntax error. Do you think that complexity is worth it to avoid an extra evaluation only on the first time the mapping tries to execute the function?

rryan replied:
Storing the exception in a member variable isn't necessary -- you can pass a pointer into checkException that checkException would write to or you could make checkException return a std::tuple or std::pair. It'd be best to not have script values stored in member variables (they need to be cleared when the script engine shuts down since they point to the old one).

rryan replied:
Also, QScriptEngine::evaluate() returns the exception if one occurred: 
```
The evaluation of program can cause an exception in the engine; in this case the return value will be the exception that was thrown (typically an Error object).
```

rryan replied:
Evaluating codeSnippet can produce benign errors (as you pointed out -- for the case when the script author hasn't defined the handler yet) so I think we shouldn't be evaluating codeSnippet here because it means we will cache the failure permanently. 

Instead, here's what I'm proposing:
```c++
// Do not evaluate(codeSnippet).
if (!syntaxIsValid(codeSnippet)) {
  wrappedFunction = QScriptValue();
} else {
  // build wrappedCode (elided)
  wrappedFunction = m_pEngine->evaluate(wrappedCode);
  checkException();
}
m_scriptWrappedFunctionCache[codeSnippet] = wrappedFunction;
```

If an exception occurs, it's either because:
* codeSnippet has a syntax error
* codeSnippet is a statement, not an expression (wrappedFunction evaluation will fail in that case with a syntax error).

In the former case, the script dialog will pop from the syntaxIsValid check. In the latter case the script dialog will pop from the checkException. 

The return value will be a function (the wrapper) on success and an invalid QScriptValue if either of the syntax error cases occurred. If we return a non-function then it'll be dropped by internalExecute() and we've already notified the user via script dialog so it doesn't need double-handing.

In the event of a transient/benign failure, the execute() following wrapFunctionCode() in MidiController will show a script error dialog with the transient exception. If the script fixes itself, then the next button press will succeed.

Be-ing replied:
Ah, sorry I misunderstood what you were saying before. I'll take a look at the error messages from evaluating the wrapper in case the codeSnippet triggers an exception.

Be-ing replied:
> If we return a non-function then it'll be dropped by internalExecute() and we've already notified the user via script dialog so it doesn't need double-handing.

I don't like having silent failure after the first occurrence. I think it's helpful to have internalExecute() print the error from the exception. If it only comes up the first time a control is used, it would be easy to hit the Close button on the dialog without taking a careful look and have the error message get buried in the log.

rryan commented on src/controllers/controllerengine.cpp:
```diff
@@ -98,30 +98,48 @@ void ControllerEngine::callFunctionOnObjects(QList<QString> scriptFunctionPrefix
+        QScriptValue function = m_pEngine->evaluate(codeSnippet);
+        if (!syntaxIsValid(codeSnippet)) {
+            wrappedFunction = m_pEngine->evaluate(
+                "(function () { print('Syntax error in: "
```
If we print "Syntax error" from the handler we produce here that means that it will get printed on every button press (not just the first) after we've already delivered a popup dialog to the user telling them that stuff is going wrong with their scripts.

If this is a script in the wild that's misbehaving this will result in spamming the user's log (which they probably aren't looking at it if they aren't technical). If this were say a jog wheel handler, that means we'd print an error message for every tick of the jog wheel. Printing too many error messages can actually slow down mixxx via priority inversion since any other thread that prints has to fight for a mutex to write to stdout.

Be-ing replied:
The reason I put the `print`s in these functions was for mapping development. The way I had it before, the popup dialog would appear on the first button press, but every press after that would be a silent failure. That could be difficult to debug.

rryan replied:
Hm ok. It seems to me the popup with syntax check failure would be enough of a warning.

If the wrapper throws an exception when it's called you'll get a script error dialog every time it fails (i.e. if you're changing around the objects the expression refers to) so those kinds of failures won't go silent.

Be-ing replied:
Yes, but getting the QScriptValue Error returned would require a going down a rabbit hole of refactoring that I spent most of yesterday puzzling about. checkException() returns a boolean and the functions that call it also generally return booleans, depending on what checkException returns. `m_pEngine->clearExceptions()` has to be called before evaluating the next code snippet, otherwise strange things happen. Without clearing the exceptions, after pressing a button that triggered an exception, the same exception would keep appearing until I used 2-3 other, valid controls but the script functions mapped to those controls would execute fine. Clearing the exceptions from checkExceptions() and syntaxIsValid() is the easiest way to do that, but then the functions that call those don't have access to the QScriptValue Error object. Also, internalExecute() then needs to change to handle the Error objects or it will just generically say they're not a function. Wrapping the error messages in the JS functions returned from wrapFunctionCode() is the most elegant way I came up with to deal with this.

Be-ing replied:
I suppose I could add a new private member to ControllerEngine to store the QScriptValue Error rather than trying to return it from functions. WDYT?

rryan commented on src/controllers/controllerengine.cpp:
```diff
@@ -98,30 +98,48 @@ void ControllerEngine::callFunctionOnObjects(QList<QString> scriptFunctionPrefix
+                "(function () { print('Syntax error in: "
+                + codeSnippet + "'); })");
+        } else if (checkException()) {
+            wrappedFunction = m_pEngine->evaluate(
```
this evaluate could fail if codeSnippet has single quotes in it

Be-ing replied:
augh. This is why I think Mixxx should stay away from dealing with strings of JS as much as possible.

rryan replied:
Carefully quarantining your untrusted data also works :).

Be-ing committed:
return QScriptValue Error objects from ControllerEngine::wrapFunctionCode()

instead of returning a JS function that prints an error. This avoids having to deal with escaping JS within JS within C++.

Be-ing commented:
Okay, I found a way around having to deal with escaping JS within JS within C++. This still gives meaningful error messages on every button press in the log, but only pops up an attention-grabbing dialog the first time the button is pressed.

Be-ing commented on src/controllers/controllerengine.cpp:
```diff
@@ -390,6 +394,12 @@ bool ControllerEngine::internalExecute(QScriptValue thisObject, QScriptValue fun
 
+    if (functionObject.isError()) {
+        qDebug() << "ControllerEngine::internalExecute:"
+                 << functionObject.toString();
```
It would be easy to call scriptErrorDialog() here, but IMO popping up a dialog repeatedly is annoying. When debugging a mapping, I'd rather look through the log than have to click to close a bunch of dialogs.

rryan commented:
> Full backtrace attached
> test-crash.txt

Thanks -- that's an odd one:
```
Thread 1 (Thread 0x7ffff7f798c0 (LWP 19496)):
#0  0x00007ffff47805da in QString::fromLocal8Bit(char const*, int) (str=0x21 <error: Cannot access memory at address 0x21>, size=size@entry=-1)
    at tools/qstring.cpp:3960
#1  0x00007ffff48466a1 in QCoreApplication::arguments() () at kernel/qcoreapplication.cpp:2345
```

I think this is probably related to google-test parsing of commandline flags (it mutates argc) and QApplication's use of command line flags. Can you try messing with this line:
https://github.com/mixxxdj/mixxx/blob/master/src/test/main.cpp#L27

For example:
* printing out the argv array before and after testing::InitGoogleTest
* passing 1 as the argc to applicationScope

Not really sure what's going on. Maybe google test is dropping argc to 0. QApplication says argc needs to be at least 1. I tried on Mac and Linux and argc is always 1 when I invoke mixxx-test as "./mixxx-test".

Be-ing committed:
only evaluate codeSnippet once in ControllerEngine::wrapFunctionCode

Be-ing commented:
Actually, I don't think wrapFunctionCode() needs to check syntax. The "Parse error" message is generic, but MidiController::processInputMapping prints the offending code snippet in the log with each button press.

Be-ing commented:
The exception error messages from evaluating the wrapper look pretty much the same as those for just executing codeSnippet, so yes, it is fine to just evaluate the wrapper.

rryan commented:
> Actually, I don't think wrapFunctionCode() needs to check syntax. The "Parse error" message is generic, but MidiController::processInputMapping prints the offending code snippet in the log with each button press.

Cool, ok -- this LGTM. Thanks for the changes!

Be-ing commented:
Great! I'll add a couple more things to the P32 mapping not depending on this that will be compatible with 2.0, then I'd like to branch off from that and take advantage of `this`.

rryan commented:
> Not really sure what's going on. Maybe google test is dropping argc to 0. QApplication says argc needs to be at least 1. I tried on Mac and Linux and argc is always 1 when I invoke mixxx-test as "./mixxx-test".

I think I fixed the problem. Uggggggh.
https://github.com/mixxxdj/mixxx/commit/8e905c7bc2f4e63b5deb42c307c2ed2020f184b9
